rep-2011.rst

```diff
@@ -0,0 +1,1314 @@
+
+This recommendation is particularly useful in the case where a DDS based ROS 2 endpoint is talking to another DDS-XTypes based ROS 2 endpoint.
+The DDS-XTypes based endpoint then has a chance to "gracefully degrade" to interoperate with the basic DDS based ROS 2 endpoint.
+This would not be the case if the DDS-XTypes based ROS 2 endpoint did not include the type version hash in the type name, as is suggested with the recommended strategy.
+
+.. TODO::
+
+    We need to confirm whether or not having a different type name prevents DDS-XTypes from working properly.
+    We've done some experiments, but we need to summarize the results and confirm the recommendation in the REP specifically geared towards DDS.
+
+Notifying the ROS Client Library
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+One potential downside to delegating type matching to the rmw implementation is that detecting the mismatch is more complicated.
+If ROS 2 is to provide users a warning that two endpoints will not communicate due to their types not matching, it requires there to be a way for the middleware to notify the ROS layer when a topic is not matched due to the type incompatibility.
+As some of the following sections describe, it might be that the rules by which the middleware decides on type compatibility are unknown to ROS 2, and so the middleware has to indicate when matches are and are not made.
+
+If the middleware just uses the type name to determine compatibility, then the rmw implementation can just check the type version hash, and if they do not match between endpoints then the rmw implementation can notify ROS 2, and a warning can be produced.
+
+Either way, to facilitate this notice, the ``rmw_event_type_t`` shall be extended to include a new event called ``RMW_EVENT_OFFERED_TYPE_INCOMPATIBLE``.
+Related functions and structures will also be updated so that the event can be associated with specific endpoints.
+
+.. TODO::
+
+    It's not clear how we will do this just now, since existing "QoS events" lack a way to communicate this information, I (wjwwood) think.
+
+Accessing the Type Version Hash
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+For debugging and introspection, the type version hash will be accessible via the ROS graph API, by extending the ``rmw_topic_endpoint_info_t`` struct, and related types and functions, to include the type version hash, ``topic_type_version_hash``, as a string.
+It should be along side the ``topic_type`` string in that struct, but the ``topic_type`` field should not include the concatenated type version hash, even if the recommended approach is used.
+Instead the type name and version hash should be separated and placed in the fields separately.
+
+This information should be transmitted as part of the discovery process.
```

Comment:
I'm not sure if I understand properly the meaning of the "discovery process" in here.

To my understanding, generally, for most DDS implementations, discovery traffic is sent prior/outside of the  security plugins domain (i.e. this information can be used for reconnaissance). If this information is sent also as part of the DDS discovery process, won't it mean that we'd be conveying it in an uncrypted manner (regardless of whether security is enabled or not in the graph)?

Also, (a final answer would of course depend on the implementation but) can you briefly comment on what's the expectation regarding additional traffic generated due to sending this information as part of discovery? Will it be just the type version hash for each ROS interface?

Reply:
Well, topic names are already sent as part of the `rmw_topic_endpoint_info_t` struct. Adding a hash in my opinion is not much of a security risk, since the hash would be degenerate, and by design, meant to not be reversible.

Furthermore, depending on how the hash is generated (e.g. factoring in the field names in the type description), it can be made more robust to rainbow table attacks (definitely more than something like, say, brute forcing of MD5 hashes for passwords.)

---

Regarding additional traffic, I'd think it'd just be the type version hash. The rest of the stuff (e.g. type description distribution) is planned to be sent after discovery, after the participants have determined that they're mismatched via the hash on discovery.

Reply:
@methylDragon is right, it should just be the version hash, which would be in addition to the type name which already being sent via discovery. So the impact to information leak and extra bandwidth used should be minimal. The hash is going to be a sha-1 of the type description struct (most likely) so that is reversible, but not easily and not for much gain I think.

The actual full type descriptions (and other meta data) would be sent via the `~/_get_type_description` service which will have to be considered in security set ups just all the other built-in node topics/services must.